### PR TITLE
FP-3145: Pull providers from GitHub releases

### PR DESCRIPTION
**Pull request by arendjr**
This removes the providers that are checked-in into the FPD repository and pulls them from the latest GitHub release instease.

The default `fpd pull` command will also pull the latest release now, instead of taking the latest blobs from `main`. There is also a `--branch` argument that allows you to pull artifacts belonging to a specific branch instead of a release. This requires a GitHub token.

There's also an `fpd build-providers` command now, which is useful for local development.

### CODE REVIEW

File: .github/workflows/reusable_build_images.yml [hunk 1]
```diff
@@ -215,14 +215,20 @@ jobs:
...
           path: artifacts
 
+      - name: Pull latest providers
+        run: cargo run pull
+        env:
+          GITHUB_TOKEN: ${{ secrets.PRIVATE_GITHUB_TOKEN }}
+
       - name: Create distributable zip file
         run: |
+          wasm_dir=$(cargo run config paths wasm-dir)
```
Comment (gagbo) [inline .github/workflows/reusable_build_images.yml:223]: Not sure, but you might need the separator here
```suggestion
          wasm_dir=$(cargo run -- config paths wasm-dir)
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -92,6 +92,11 @@ jobs:
           token: ${{ secrets.GITHUB_TOKEN }}
           args: --all-features
 
+      - name: Pull latest providers
+        run: cargo run pull
```
Comment (gagbo) [inline .github/workflows/build.yml]: Not sure, but you might need the separator here

```suggestion
        run: cargo run -- pull
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -92,6 +92,11 @@ jobs:
           token: ${{ secrets.GITHUB_TOKEN }}
           args: --all-features
 
+      - name: Pull latest providers
+        run: cargo run pull --branch=main
```
Comment (gagbo) [inline .github/workflows/build.yml:96]: Not sure, but you might need the separator there

```suggestion
        run: cargo run -- pull --branch=main
```

File: .github/workflows/reusable_build_images.yml [hunk 4]
```diff
@@ -215,14 +215,18 @@ jobs:
         with:
           path: artifacts
 
+      - name: Pull latest providers
+        run: cargo run pull
```
Comment (gagbo) [inline .github/workflows/reusable_build_images.yml:219]: Not sure, but you might need the separator there

```suggestion
        run: cargo run -- pull
```

File: src/tasks/provider_manager.rs [hunk 5]
```diff
@@ -50,41 +91,253 @@ pub async fn pull(providers: &[BuiltinProvider], all: bool) -> Result<(), Error>
...
+    providers_dir: PathBuf,
+}
+
+pub fn build_providers(args: BuildProvidersArgs) -> Result<(), Error> {
+    ensure_wasm_dir()?;
+
+    let providers_dir = args.providers_dir;
+    if !providers_dir.exists() {
+        return Err(Error::Other {
+            message: format!("Providers dir `{}` doesn't exist", providers_dir.display()),
```
Comment (gagbo) [inline src/tasks/provider_manager.rs:253]: I don't remember if I left that message like this in the beginning, but I think that error messages that eventually will be read by users should avoid abbreviations when possible

```suggestion
            message: format!("Providers directory `{}` doesn't exist", providers_dir.display()),
```

File: .github/workflows/reusable_build_images.yml (hunk 1 above)
Comment (arendjr) [reply .github/workflows/reusable_build_images.yml:223]: Nah, that's only necessary if you want to pass options directly after `cargo run`. If you pass another command it understands you've moved on to the arguments of the command being run.

**Review by gagbo — APPROVED** [review]

**Pull request merged**